File: apps/gig-crawler-2/src/adapters/BraveSearchRepo/BraveSearchAdapter.ts
import axios from "axios";
import type { SearchPort } from "../../ports/SearchPort.js";
import type { SearchResult } from "../../models/searchResult.js";
import { logger } from "../../utils/logger.js";
import { retry } from "../../utils/retry.js";
import { env } from "../../models/env.js";

export class BraveSearchAdapter implements SearchPort {
  private readonly baseUrl = "https://api.search.brave.com/res/v1/web/search";
  private readonly apiKey: string;

  constructor(apiKey: string = env.BRAVE_API_KEY) {
    this.apiKey = apiKey;
  }

  async search(query: string, count: number = 20): Promise<SearchResult[]> {
    return retry(
      async () => {
        logger.info({ query, count }, "Searching Brave");

        const response = await axios.get(this.baseUrl, {
          headers: {
            Accept: "application/json",
            "X-Subscription-Token": this.apiKey,
          },
          params: {
            q: query,
            count: Math.min(count, 20),
          },
        });

        const results: SearchResult[] = [];
        const webResults = response.data.web?.results || [];

        for (const result of webResults) {
          results.push({
            url: result.url,
            title: result.title,
            description: result.description,
          });
        }

        logger.info({ count: results.length }, "Found search results");
        return results;
      },
      {
        maxAttempts: 3,
        onError: (error, attempt) => {
          logger.warn({ error, attempt }, "Brave search attempt failed");
        },
      }
    );
  }
}

File: apps/gig-crawler-2-python-backup/src/adapters/brave_search/brave_search_adapter.py
"""Brave Web Search API adapter."""
import httpx
from typing import List
from ...ports.search_port import SearchPort
from ...models.search_result import SearchResult
from ...utils.logger import setup_logger
from ...utils.retry import async_retry

logger = setup_logger(__name__)


class BraveSearchAdapter(SearchPort):
    """Brave Web Search API implementation."""

    BASE_URL = "https://api.search.brave.com/res/v1/web/search"

    def __init__(self, api_key: str):
        """Initialize with API key.

        Args:
            api_key: Brave API key
        """
        self.api_key = api_key
        self.client = httpx.AsyncClient(timeout=30.0)

    async def close(self):
        """Close HTTP client."""
        await self.client.aclose()

    @async_retry(max_attempts=3, exceptions=(httpx.HTTPError,))
    async def search(self, query: str, count: int = 20) -> List[SearchResult]:
        """Search the web using Brave Web Search API.

        Args:
            query: Search query string
            count: Number of results to return (max 20)

        Returns:
            List of search results
        """
        logger.info(f"Searching Brave for: {query}")

        headers = {
            "Accept": "application/json",
            "X-Subscription-Token": self.api_key,
        }

        params = {
            "q": query,
            "count": min(count, 20),
        }

        response = await self.client.get(
            self.BASE_URL,
            headers=headers,
            params=params,
        )
        response.raise_for_status()

        data = response.json()
        results = []

        # Parse web results
        web_results = data.get("web", {}).get("results", [])
        for result in web_results:
            results.append(
                SearchResult(
                    url=result.get("url", ""),
                    title=result.get("title", ""),
                    description=result.get("description"),
                )
            )

        logger.info(f"Found {len(results)} search results")
        return results
